File: crates/ui/src/badges.rs
//! Message badges: pills standing in for structured context a user message
//! carries.
//!
//! A feature stages context on the composer, folds it into the prompt as plain
//! text, and registers an [`Extractor`]. [`split`] lifts that block back out of
//! the sent message so the transcript draws a pill instead of the bullets the
//! agent reads. Nothing here knows what a diff comment is.

use std::sync::Arc;
use std::time::Duration;

use gpui::{Context, IntoElement, Render, SharedString, Window, div, prelude::*, px};

use crate::i18n::{self, Locale, MessageId};
use crate::theme::Theme;

/// What a pill says. Kept as data rather than a string because the count is
/// known when the message is split, and the phrase only when it is drawn.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum BadgeLabel {
    Comments(usize),
}

impl BadgeLabel {
    pub fn text(&self, locale: Locale) -> String {
        match self {
            BadgeLabel::Comments(count) => i18n::counted(
                MessageId::CountCommentOne,
                MessageId::CountCommentMany,
                *count,
                locale,
            ),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct MessageBadge {
    /// An `icons::*` asset name.
    pub icon: &'static str,
    pub label: BadgeLabel,
    /// Empty means the label says everything and the pill carries no card.
    pub details: Vec<BadgeDetail>,
}

/// One row of a hover card. Three generic slots, so a new badge kind fills them
/// without this module learning what it cites.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct BadgeDetail {
    /// `src/main.rs:42`, a URL, a commit subject.
    pub location: SharedString,
    /// `L`/`R` for a diff side, a status, a count.
    pub tag: Option<SharedString>,
    pub body: SharedString,
}

/// Returns the text with this feature's block removed, plus the pill replacing
/// it. `None` when the message carries nothing of that kind.
pub type Extractor = fn(&str) -> Option<(String, MessageBadge)>;

const EXTRACTORS: &[Extractor] = &[crate::comments::extract_badge];

/// Each extractor sees what the previous ones left behind, so two features can
/// ride the same prompt.
pub fn split(text: &str) -> (String, Vec<MessageBadge>) {
    let mut rest = text.to_string();
    let mut badges = Vec::new();
    for extract in EXTRACTORS {
        if let Some((stripped, badge)) = extract(&rest) {
            rest = stripped;
            badges.push(badge);
        }
    }
    (rest, badges)
}

/// The composer sizes its strip arithmetically, so this cannot be a
/// measurement.
pub const BADGE_HEIGHT: f32 = 24.0;

const PILL_RADIUS: f32 = 8.0;
const ICON_SIZE: f32 = 12.0;
const TEXT_SIZE: f32 = 12.0;
const CARD_WIDTH: f32 = 320.0;
const HOVER_DELAY: Duration = Duration::from_millis(280);

/// `id` scopes the hover card and must be stable per rendered pill.
pub fn render(
    id: impl Into<gpui::ElementId>,
    badge: &MessageBadge,
    theme: &Theme,
    locale: Locale,
) -> gpui::Stateful<gpui::Div> {
    div()
        .id(id)
        .h(px(BADGE_HEIGHT))
        .flex()
        .flex_row()
        .items_center()
        .gap(px(6.0))
        .px(px(8.0))
        .rounded(px(PILL_RADIUS))
        .bg(crate::theme::ink(0.06))
        .text_size(px(TEXT_SIZE))
        .font_weight(gpui::FontWeight::MEDIUM)
        .text_color(theme.text_muted)
        .child(
            crate::icons::icon(badge.icon)
                .size(px(ICON_SIZE))
                .text_color(theme.text_muted.opacity(0.7)),
        )
        .child(badge.label.text(locale))
        .when(!badge.details.is_empty(), |el| {
            let details = Arc::new(badge.details.clone());
            el.tooltip(move |_, cx| {
                cx.new(|_| BadgeCard {
                    details: details.clone(),
                })
                .into()
            })
            .tooltip_show_delay(HOVER_DELAY)
        })
}

struct BadgeCard {
    details: Arc<Vec<BadgeDetail>>,
}

impl BadgeCard {
    fn row(detail: &BadgeDetail, theme: &Theme) -> gpui::Div {
        div()
            .flex()
            .flex_row()
            .gap(px(8.0))
            .p(px(8.0))
            .rounded(px(6.0))
            .bg(crate::theme::ink(0.05))
            .child(
                div()
                    .flex_none()
                    .w(px(2.0))
                    .rounded(px(1.0))
                    .bg(theme.solid.opacity(0.35)),
            )
            .child(
                div()
                    .flex_1()
                    .min_w_0()
                    .flex()
                    .flex_col()
                    .gap(px(4.0))
                    .child(
                        div()
                            .flex()
                            .flex_row()
                            .items_center()
                            .gap(px(6.0))
                            .font_family(theme.font_mono.clone())
                            .text_size(px(10.0))
                            .text_color(theme.text_faint)
                            .child(
                                div()
                                    .flex_1()
                                    .min_w_0()
                                    .truncate()
                                    .child(detail.location.clone()),
                            )
                            .children(detail.tag.clone().map(|tag| {
                                div()
                                    .flex_none()
                                    .px(px(4.0))
                                    .rounded(px(3.0))
                                    .bg(crate::theme::ink(0.10))
                                    .child(tag)
                            })),
                    )
                    .child(
                        div()
                            .min_w_0()
                            .text_size(px(TEXT_SIZE))
                            .line_height(px(16.0))
                            .text_color(theme.text)
                            .child(detail.body.clone()),
                    ),
            )
    }
}

impl Render for BadgeCard {
    fn render(&mut self, _window: &mut Window, cx: &mut Context<Self>) -> impl IntoElement {
        let theme = &Theme::of(cx).for_popup();
        let card = crate::popover::popover_card(theme)
            .w(px(CARD_WIDTH))
            .p(px(6.0))
            .flex()
            .flex_col()
            .gap(px(4.0))
            .children(self.details.iter().map(|detail| Self::row(detail, theme)));
        // `frosted` is load-bearing beyond the blur: it gives the card its own
        // scene layer. Sharing the transcript's layer let the bubble's text
        // paint over the card, since equal draw orders render quads before
        // text.
        crate::frost::frosted(crate::popover::CARD_RADIUS, crate::frost::MENU_BLUR, card)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::comments::{CommentSide, ReviewComment, with_comments};

    #[test]
    fn a_plain_message_carries_no_badges() {
        let (text, badges) = split("just a prompt");
        assert_eq!(text, "just a prompt");
        assert!(badges.is_empty());
    }

    #[test]
    fn a_sent_comment_block_becomes_one_pill() {
        let staged = vec![
            ReviewComment::new("a.rs", CommentSide::New, 3, "fix"),
            ReviewComment::new("b.rs", CommentSide::Old, 9, "why"),
        ];
        let (text, badges) = split(&with_comments("do", &staged));
        assert_eq!(text, "do");
        assert_eq!(badges.len(), 1);
        assert_eq!(badges[0].label.text(Locale::En), "2 comments");
        assert_eq!(badges[0].label.text(Locale::ZhCn), "2 条评论");
    }

    #[test]
    fn a_comment_count_pluralizes_in_both_locales() {
        let counts = [(1, "1 comment"), (2, "2 comments"), (0, "0 comments")];
        for (count, expected) in counts {
            assert_eq!(BadgeLabel::Comments(count).text(Locale::En), expected);
        }
        assert_eq!(BadgeLabel::Comments(1).text(Locale::ZhCn), "1 条评论");
    }

    #[test]
    fn the_card_carries_one_row_per_comment() {
        let staged = vec![
            ReviewComment::new("src/main.rs", CommentSide::New, 42, "early-return here"),
            ReviewComment::new("src/lib.rs", CommentSide::Old, 7, "why was this dropped?"),
        ];
        let (_, badges) = split(&with_comments("look", &staged));
        let details = &badges[0].details;
        assert_eq!(details.len(), 2);
        assert_eq!(details[0].location.as_ref(), "src/main.rs:42");
        assert_eq!(details[0].tag.as_deref(), Some("R"));
        assert_eq!(details[0].body.as_ref(), "early-return here");
        assert_eq!(details[1].location.as_ref(), "src/lib.rs:7");
        assert_eq!(details[1].tag.as_deref(), Some("L"));
    }

    #[test]
    fn a_multiline_body_rejoins_its_continuation_lines() {
        let staged = vec![ReviewComment::new(
            "a.rs",
            CommentSide::New,
            3,
            "first\nsecond",
        )];
        let (_, badges) = split(&with_comments("x", &staged));
        assert_eq!(badges[0].details[0].body.as_ref(), "first\nsecond");
    }

    #[test]
    fn a_path_with_a_colon_still_splits_on_the_side_marker() {
        let staged = vec![ReviewComment::new("odd:name.rs", CommentSide::New, 5, "hm")];
        let (_, badges) = split(&with_comments("x", &staged));
        assert_eq!(badges[0].details[0].location.as_ref(), "odd:name.rs:5");
        assert_eq!(badges[0].details[0].body.as_ref(), "hm");
    }

    #[test]
    fn a_comment_only_send_keeps_its_stand_in_body() {
        let staged = vec![ReviewComment::new("a.rs", CommentSide::New, 3, "fix")];
        let (text, badges) = split(&with_comments("", &staged));
        assert_eq!(text, crate::comments::COMMENT_ONLY_TEXT);
        assert_eq!(badges[0].label.text(Locale::En), "1 comment");
    }
}

File: web/packages/app/src/lib/badges.ts
/**
 * Message badges — `crates/ui/src/badges.rs` + the one extractor,
 * `crates/ui/src/comments.rs::extract_badge` (:158-238), ported 1:1.
 *
 * A feature stages context on the composer, folds it into the prompt as plain
 * text, and registers an [`Extractor`]. [`splitBadges`] lifts that block back
 * out of the sent message so the transcript draws a pill instead of the
 * bullets the agent reads. Nothing here knows what a diff comment is.
 */

import type { IconName } from "@zeron/icons";

/** A pill standing in for structured context a user message carries. */
export interface MessageBadge {
  /** `icons::*` asset name (`chatRoundLine` for the comment badge). */
  readonly icon: IconName;
  readonly label: string;
  /** Empty means the label says everything and the pill carries no card. */
  readonly details: readonly BadgeDetail[];
}

/** One row of a hover card — three generic slots (badges.rs:26-34). */
export interface BadgeDetail {
  /** `src/main.rs:42`, a URL, a commit subject. */
  readonly location: string;
  /** `L`/`R` for a diff side, a status, a count. */
  readonly tag: string | null;
  readonly body: string;
}

/**
 * Returns the text with this feature's block removed, plus the pill replacing
 * it. `null` when the message carries nothing of that kind (badges.rs:38).
 */
export type Extractor = (text: string) => { readonly text: string; readonly badge: MessageBadge } | null;

/** `COMMENT_BLOCK_HEADER` (comments.rs:123) — every comment is a diff comment. */
export const COMMENT_BLOCK_HEADER =
  "Comments on the diff (each cites the file and line it belongs to; L = line number in the original file, R = in the changed file):";
/** `REVIEW_COMMENT_BLOCK_HEADER` (comments.rs:124-125) — any file comment rides this one. */
export const REVIEW_COMMENT_BLOCK_HEADER =
  "Review comments (each cites the workspace file and line it belongs to):";
/** The stand-in body a comment-only send carries (comments.rs:121). */
export const COMMENT_ONLY_TEXT = "Address the review comments below.";

/**
 * Each extractor sees what the previous ones left behind, so two features can
 * ride the same prompt (badges.rs:44-54). The array is the extension point —
 * a future extractor just appends.
 */
export function splitBadges(text: string): { readonly text: string; readonly badges: readonly MessageBadge[] } {
  let rest = text;
  const badges: MessageBadge[] = [];
  for (const extract of EXTRACTORS) {
    const out = extract(rest);
    if (out !== null) {
      rest = out.text;
      badges.push(out.badge);
    }
  }
  return { text: rest, badges };
}

/**
 * `comments::extract_badge` (comments.rs:158-189): the comment block, matched
 * only as a whole TRAILING block, so a prompt quoting the header mid-body is
 * left alone. Accepts either header; the LAST occurrence wins.
 */
export function extractCommentBadge(text: string): { readonly text: string; readonly badge: MessageBadge } | null {
  const markers = [`\n\n${COMMENT_BLOCK_HEADER}\n`, `\n\n${REVIEW_COMMENT_BLOCK_HEADER}\n`];
  let at = -1;
  let marker = "";
  for (const candidate of markers) {
    const found = text.lastIndexOf(candidate);
    if (found > at) {
      at = found;
      marker = candidate;
    }
  }
  if (at < 0) {
    return null;
  }
  const block = text.slice(at + marker.length);
  if (
    block.length === 0 ||
    !block
      .split("\n")
      .every((line) => line.startsWith("- ") || line.startsWith("  "))
  ) {
    return null;
  }
  const details = parseBullets(block);
  if (details.length === 0) {
    return null;
  }
  return {
    text: text.slice(0, at),
    badge: {
      icon: "chatRoundLine",
      label: chipLabel(details.length),
      details,
    },
  };
}

/** The extractor registry (badges.rs:40) — one entry today. */
const EXTRACTORS: readonly Extractor[] = [extractCommentBadge];

/**
 * `parse_bullets` (comments.rs:190-238): one `BadgeDetail` per `"- "` bullet;
 * a `"  "` continuation line appends to the previous detail's body (that is
 * how a multi-line comment survives the round trip). For a bullet, two
 * candidate splits are computed and the EARLIEST wins — a body may itself
 * contain `"(L): "`, and matching that would swallow the body into the
 * location.
 */
export function parseBullets(block: string): BadgeDetail[] {
  const details: BadgeDetail[] = [];
  for (const line of block.split("\n")) {
    if (!line.startsWith("- ")) {
      const indented = line.startsWith("  ") ? line.slice(2) : null;
      const last = details[details.length - 1];
      if (indented !== null && last !== undefined) {
        details[details.length - 1] = { ...last, body: `${last.body}\n${indented}` };
      }
      continue;
    }
    const bullet = line.slice(2);
    // Side markers `" (L): "` / `" (R): "` (comments.rs:13-19, :127-129).
    let side: { at: number; markerLen: number; tag: string } | null = null;
    for (const tag of ["L", "R"]) {
      const marker = ` (${tag}): `;
      const at = bullet.indexOf(marker);
      if (at >= 0 && (side === null || at < side.at)) {
        side = { at, markerLen: marker.length, tag };
      }
    }
    const fileSplit = splitFileBullet(bullet);
    if (fileSplit !== null && (side === null || fileSplit.at < side.at)) {
      details.push({ location: fileSplit.location, tag: null, body: fileSplit.body });
      continue;
    }
    if (side !== null) {
      details.push({
        location: bullet.slice(0, side.at),
        tag: side.tag,
        body: bullet.slice(side.at + side.markerLen),
      });
      continue;
    }
    if (fileSplit !== null) {
      details.push({ location: fileSplit.location, tag: null, body: fileSplit.body });
    }
  }
  return details;
}

/**
 * `split_file_bullet` (comments.rs:239-246): the first `": "` whose preceding
 * text ends in `:{digits}` — i.e. `"{path}:{line}"` parses as a line number.
 */
function splitFileBullet(bullet: string): { at: number; location: string; body: string } | null {
  let search = 0;
  for (;;) {
    const at = bullet.indexOf(": ", search);
    if (at < 0) {
      return null;
    }
    search = at + 1;
    const location = bullet.slice(0, at);
    const colon = location.lastIndexOf(":");
    if (colon < 0) {
      continue;
    }
    const line = location.slice(colon + 1);
    if (/^\d+$/.test(line)) {
      return { at, location, body: bullet.slice(at + 2) };
    }
  }
}

/** `chip_label` (comments.rs:247-253). */
export function chipLabel(count: number): string {
  return count === 1 ? "1 comment" : `${count} comments`;
}
